File: app/schemas/governance.py
"""
Governance Schemas
Includes Resource Ownership Transfers and System Configuration Schemas.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field
from app.models.governance import (
    ConfigValueType,
    TransferResourceType,
    TransferStatus,
)


# -----------------------------------------------------------------------------
# Ownership Transfer & Account Succession Schemas
# -----------------------------------------------------------------------------

class OwnershipTransferCreate(BaseModel):
    resource_type: TransferResourceType
    resource_id: UUID
    requested_owner_id: UUID
    reason: str = Field(..., min_length=5)


class SuccessionUserSummary(BaseModel):
    id: UUID
    username: str
    full_name: str
    email: Optional[str] = None
    account_status: str
    role_name: Optional[str] = None


class SuccessionTaskSummary(BaseModel):
    id: UUID
    title: str
    priority: str
    status: str
    vertical_name: Optional[str] = None


class SuccessionEventSummary(BaseModel):
    id: UUID
    name: str
    status: str
    role: str


class SuccessionVerticalSummary(BaseModel):
    id: UUID
    name: str
    is_primary: bool


class AccountSuccessionPreviewResponse(BaseModel):
    previous_user: SuccessionUserSummary
    successor_user: SuccessionUserSummary
    active_tasks_count: int
    active_tasks: List[SuccessionTaskSummary] = Field(default_factory=list)
    active_events_count: int
    active_events: List[SuccessionEventSummary] = Field(default_factory=list)
    active_requirements_count: int
    assigned_verticals: List[SuccessionVerticalSummary] = Field(default_factory=list)
    historical_preservation_note: str = (
        "Historical completed tasks, submitted reports, meetings, past communications, "
        "and audit logs will remain permanently preserved under the Previous Account."
    )


class AccountSuccessionCreate(BaseModel):
    previous_user_id: UUID
    successor_user_id: UUID
    reason: str = Field(..., min_length=5, description="Administrative justification for account succession")


class OwnershipTransferReviewRequest(BaseModel):
    status: TransferStatus = Field(..., description="Must be APPROVED or REJECTED")
    remarks: Optional[str] = None


class OwnershipTransferResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    resource_type: TransferResourceType
    resource_id: UUID
    resource_name: Optional[str] = None
    current_owner_id: UUID
    current_owner_username: Optional[str] = None
    requested_owner_id: UUID
    requested_owner_username: Optional[str] = None
    requested_by_id: UUID
    requested_by_username: Optional[str] = None
    reviewed_by_id: Optional[UUID] = None
    reviewed_by_username: Optional[str] = None
    reason: str
    status: TransferStatus
    remarks: Optional[str] = None
    created_at: datetime
    reviewed_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class OwnershipTransferListResponse(BaseModel):
    total: int
    items: List[OwnershipTransferResponse]


# -----------------------------------------------------------------------------
# System Configuration Schemas
# -----------------------------------------------------------------------------

class SystemConfigCreate(BaseModel):
    key: str = Field(..., min_length=2, max_length=100)
    value: str = Field(...)
    value_type: ConfigValueType = ConfigValueType.STRING
    description: Optional[str] = Field(None, max_length=255)
    is_active: bool = True


class SystemConfigUpdate(BaseModel):
    value: str = Field(...)
    description: Optional[str] = Field(None, max_length=255)
    is_active: Optional[bool] = None


class SystemConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    key: str
    value: str
    value_type: ConfigValueType
    description: Optional[str] = None
    is_active: bool
    updated_by_id: Optional[UUID] = None
    updated_by_username: Optional[str] = None
    updated_at: datetime


class SystemConfigListResponse(BaseModel):
    total: int
    items: List[SystemConfigResponse]

File: frontend/types/governance.ts
/**
 * Governance Domain Types
 * Matches backend schemas in app/schemas/governance.py and app/schemas/audit.py
 */

export type TransferResourceType = 'ACCOUNT' | 'TASK' | 'EVENT' | 'REQUIREMENT';
export type TransferStatus = 'PENDING' | 'APPROVED' | 'REJECTED' | 'CANCELLED' | 'COMPLETED';

export type ConfigValueType = 'STRING' | 'INTEGER' | 'FLOAT' | 'BOOLEAN' | 'JSON';

export interface OwnershipTransferCreate {
  resource_type: TransferResourceType;
  resource_id: string;
  requested_owner_id: string;
  reason: string;
}

export interface SuccessionUserSummary {
  id: string;
  username: string;
  full_name: string;
  email?: string | null;
  account_status: string;
  role_name?: string | null;
}

export interface SuccessionTaskSummary {
  id: string;
  title: string;
  priority: string;
  status: string;
  vertical_name?: string | null;
}

export interface SuccessionEventSummary {
  id: string;
  name: string;
  status: string;
  role: string;
}

export interface SuccessionVerticalSummary {
  id: string;
  name: string;
  is_primary: boolean;
}

export interface AccountSuccessionPreviewResponse {
  previous_user: SuccessionUserSummary;
  successor_user: SuccessionUserSummary;
  active_tasks_count: number;
  active_tasks: SuccessionTaskSummary[];
  active_events_count: number;
  active_events: SuccessionEventSummary[];
  active_requirements_count: number;
  assigned_verticals: SuccessionVerticalSummary[];
  historical_preservation_note: string;
}

export interface AccountSuccessionCreate {
  previous_user_id: string;
  successor_user_id: string;
  reason: string;
}

export interface OwnershipTransferReviewRequest {
  status: TransferStatus;
  remarks?: string;
}

export interface OwnershipTransferResponse {
  id: string;
  resource_type: TransferResourceType;
  resource_id: string;
  resource_name?: string | null;
  current_owner_id: string;
  current_owner_username?: string | null;
  requested_owner_id: string;
  requested_owner_username?: string | null;
  requested_by_id: string;
  requested_by_username?: string | null;
  reviewed_by_id?: string | null;
  reviewed_by_username?: string | null;
  reason: string;
  status: TransferStatus;
  remarks?: string | null;
  created_at: string;
  reviewed_at?: string | null;
  completed_at?: string | null;
}

export interface OwnershipTransferListResponse {
  total: number;
  items: OwnershipTransferResponse[];
}

export interface SystemConfigCreate {
  key: string;
  value: string;
  value_type?: ConfigValueType;
  description?: string;
  is_active?: boolean;
}

export interface SystemConfigUpdate {
  value: string;
  description?: string;
  is_active?: boolean;
}

export interface SystemConfigResponse {
  id: string;
  key: string;
  value: string;
  value_type: ConfigValueType;
  description?: string | null;
  is_active: boolean;
  updated_by_id?: string | null;
  updated_by_username?: string | null;
  updated_at: string;
}

export interface SystemConfigListResponse {
  total: number;
  items: SystemConfigResponse[];
}

export interface AuditLogResponse {
  id: string;
  timestamp: string;
  action: string;
  resource_type: string;
  resource_id?: string | null;
  actor_id?: string | null;
  actor_username?: string | null;
  correlation_id?: string | null;
  ip_address?: string | null;
  outcome: string;
  details?: Record<string, unknown> | null;
  created_at?: string;
}

export interface AuditLogListResponse {
  total: number;
  items: AuditLogResponse[];
}
